### PR TITLE
Basic stamping support

### PR DESCRIPTION
**Pull request by alexeagle**

### CODE REVIEW

File: lib/stamping.bzl [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+def stamp_files(ctx):
+    """Provide the files that contain the result of the --workspace_status_command.
+
+    Returns:
+        If stamping is not enabled for the current build, returns None.
+        Otherwise, returns a struct containing (volatile_status, stable_status) keys
+    """
+    if is_stamping_enabled(ctx.attr):
+        return struct(
+            volatile_status = ctx.version_file,
```
Comment (gregmagolan) [inline lib/stamping.bzl:99]: `volatile_status_file` and `stable_status_file` as the keys? the current values make downstream harder to read as `_status` implies some status file and not a file.
```
if stamp:
        args.add("--volatile_status_file", stamp.volatile_status_file)
        args.add("--stable_status_file", stamp.stable_status_file)

        inputs.extend([stamp.volatile_status_file, stamp.stable_status_file])
```

Comment (gregmagolan) [reply lib/stamping.bzl:99]: Funny how version_file and info_file are terribly non-descriptive so we rename them here

**Comment by gregmagolan** [comment]
Add a 2nd commit to use this in run_binary?

File: lib/stamping.bzl [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+)
+```
+"""
+
+load("//lib/private:stamping.bzl", "is_stamping_enabled")
+
+def stamp_files(ctx):
+    """Provide the files that contain the result of the --workspace_status_command.
+
+    Returns:
```
Comment (gregmagolan) [inline lib/stamping.bzl:93]:
```
Args:
    ctx: The rule context
```

File: lib/stamping.bzl [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+        If stamping is not enabled for the current build, returns None.
```
Comment (gregmagolan) [inline lib/stamping.bzl:94]: "If stamping is not enabled for rule in the current build, returns None."

File: lib/stamping.bzl [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+        # ... my attributes ...
+    }, **STAMP_ATTRS),
+)
+```
+"""
+
+load("//lib/private:stamping.bzl", "is_stamping_enabled")
+
+def stamp_files(ctx):
+    """Provide the files that contain the result of the --workspace_status_command.
```
Comment (gregmagolan) [inline lib/stamping.bzl:91]: You still get the files even if you don't specify a `--workspace_status_command` right? Just with the default values Bazel provides

File: lib/stamping.bzl (hunk 3 above)
Comment (gregmagolan) [reply lib/stamping.bzl:94]: Perhaps, elaborate here that stamping will be turned on for the rule if `stamp = 1` or if `stamp = -1` and the `--stamp` config flag is set

File: lib/stamping.bzl (hunk 2 above)
Comment (alexeagle) [reply lib/stamping.bzl:93]: buildifier didn't tell me, I'm confused when those warnings appear. feels like a bug, it starts yelling at me only later...

File: lib/stamping.bzl (hunk 4 above)
Comment (alexeagle) [reply lib/stamping.bzl:91]: yes that's true

File: lib/stamping.bzl [hunk 5]
```diff
@@ -0,0 +1,128 @@
...
+    attrs = dict({
+        # ... my attributes ...
+    }, **STAMP_ATTRS),
+)
+```
+"""
+
+load("//lib/private:stamping.bzl", "is_stamping_enabled")
+
+def stamp_files(ctx):
```
Comment (gregmagolan) [inline lib/stamping.bzl:90]: Pondering a better name than `stamp_files` for this function that better describes what is does,

```
stamp = check_stamp(ctx)
```

```
stamp = stamp_info(ctx)
```

```
stamp = maybe_stamp(ctx)
```

Comment (alexeagle) [reply lib/stamping.bzl]: I like `maybe_stamp`

**Review by gregmagolan — APPROVED** [review]
🌮

**Comment by gregmagolan** [comment]
squash commit 1 through 3 into 1 and then land with the 2 commits unsquahsed?

**Comment by gregmagolan** [comment]
I already have

```
        "BAZEL_INFO_FILE": "$(INFO_FILE)",
        "BAZEL_VERSION_FILE": "$(VERSION_FILE)",
```

set always in js_run_binary. I'll remove those from there then and update to this new run_binary under the hood"

**Pull request merged**